cedar-policy-validator/src/typecheck.rs

```diff
@@ -1961,83 +1964,126 @@ impl<'a> Typechecker<'a> {
-                // Still return `TypecheckFail` so that typechecking is not considered successful.
-                Some(EntityType::Unspecified) => TypecheckAnswer::fail(
+            }
+            (Some(EntityType::Specified(_)), None) => {
+                // One or more of the elements on the right is not an entity
+                // literal. The `in` is still valid, so typechecking succeeds
+                // with type Boolean.
+                TypecheckAnswer::success(
                     ExprBuilder::with_data(Some(Type::primitive_boolean()))
                         .with_same_source_loc(in_expr)
                         .is_in(lhs_expr, rhs_expr),
-                ),
+                )
+            }
+            (Some(EntityType::Unspecified), Some(rhs)) => {
+                // It's perfectly valid for `principal` or `resource` to be `EntityType::Unspecified`
+                if rhs
+                    .iter()
+                    .any(|euid| matches!(euid.entity_type(), EntityType::Unspecified))
+                {
+                    // something on the RHS is unspecified, so we have to type `unspecified in RHS` as Bool,
+                    // because two unspecified entities are equal (and thus `in`) if they have the same `Eid`.
+                    TypecheckAnswer::success(
+                        ExprBuilder::with_data(Some(Type::primitive_boolean()))
+                            .with_same_source_loc(in_expr)
+                            .is_in(lhs_expr, rhs_expr),
+                    )
+                } else {
+                    // nothing on the RHS is unspecified, so `unspecified in RHS` is always false
+                    TypecheckAnswer::success(
+                        ExprBuilder::with_data(Some(Type::singleton_boolean(false)))
+                            .with_same_source_loc(in_expr)
+                            .is_in(lhs_expr, rhs_expr),
+                    )
+                }
+            }
+            (Some(EntityType::Unspecified), None) => {
+                // It's perfectly valid for `principal` or `resource` to be `EntityType::Unspecified`
+                if rhs_elems
+                    .into_iter()
+                    .any(|expr| Typechecker::is_unspecified_entity(request_env, expr))
+                {
+                    // something on the RHS is unspecified, so we have to type `unspecified in RHS` as Bool,
+                    // because two unspecified entities are equal (and thus `in`) if they have the same `Eid`.
+                    TypecheckAnswer::success(
+                        ExprBuilder::with_data(Some(Type::primitive_boolean()))
+                            .with_same_source_loc(in_expr)
+                            .is_in(lhs_expr, rhs_expr),
+                    )
+                } else {
+                    // nothing on the RHS can be unspecified, so `unspecified in RHS` is always false
+                    TypecheckAnswer::success(
```

Comment:
This might change strict validation for something like `principal in [resource.owner]` when `principal` is unspecified and `resource.owner` is some entity type.

Reply:
I think you're right.  That doesn't mean this precision increase is invalid, just that it's currently breaking (unless/until we change `ImpossiblePolicy` from error to warning)

Reply:
Actually I'm not sure if you're correct.  I believe `principal in [resource.owner]` when `principal` is unspecified and `resource.owner` is some entity type, might short-circuit out [here](https://github.com/cedar-policy/cedar/blob/2c456758391cfff0eb23d79d4c9d446ce09fd035/cedar-policy-validator/src/typecheck.rs#L1763) and never reach this function at all.

Reply:
Agreed that earlier handling handles this case. But I'm not sure this new definition matches the spec -- the Dafny code will always return `AnyBool` in the case where the RHS cannot be converted to a list of Entity UIDs.